backend/services/implementations/notificationService.ts

```diff
@@ -0,0 +1,151 @@
+  async getNotifications(
+    user: string,
+    skip: number,
+    limit: number,
+  ): Promise<NotificationsResponseDTO> {
+    let notifications: UserNotification[];
+    let totalNumberOfNotifications: number;
+    let numberOfUnseenNotifications: number;
+    try {
+      const foundNotifications = await MgNotification.find({ user })
+        .sort({ createdAt: -1 })
+        .skip(skip)
+        .limit(limit);
+
+      notifications = foundNotifications.map((notification) => {
+        return {
+          id: notification.id,
+          message: notification.message,
+          read: notification.read,
+          createdAt: notification.createdAt,
+          link: notification.link,
+        };
+      });
+
+      totalNumberOfNotifications = await MgNotification.where({
+        user,
+      }).countDocuments();
+
+      numberOfUnseenNotifications = await MgNotification.where({
+        user,
+        read: false,
+      }).countDocuments();
+    } catch (error) {
+      Logger.error(
+        `Failed to retrieve notifications. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+    return {
+      notifications,
+      totalNumberOfNotifications,
+      numberOfUnseenNotifications,
+    };
+  }
+
+  async createNotification(
+    notification: CreateNotificationDTO,
+  ): Promise<NotificationDTO> {
+    let newNotification: Notification;
+    try {
+      newNotification = await MgNotification.create({
+        user: notification.user,
+        message: notification.message,
+        link: notification.link,
+      });
+    } catch (error) {
+      Logger.error(
+        `Failed to create notification. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+    return {
+      id: newNotification.id,
+      user: newNotification.user,
+      message: newNotification.message,
+      read: newNotification.read,
+      createdAt: newNotification.createdAt,
+      link: newNotification.link,
+    };
+  }
+
+  async updateNotification(
+    notificationId: string,
+    updateNotification: UpdateNotificationDTO,
+  ): Promise<NotificationDTO> {
+    let oldNotification: Notification | null;
+    try {
+      oldNotification = await MgNotification.findByIdAndUpdate(
+        notificationId,
+        updateNotification,
+        { runValidators: true },
+      );
+      if (!oldNotification) {
+        throw new Error(`Notification ${notificationId} not found.`);
+      }
+    } catch (error) {
+      Logger.error(
+        `Failed to update notification. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+
+    // this is digustingly bad ngl
+    const message = updateNotification.message
+      ? updateNotification.message
+      : oldNotification.message;
+    const read = updateNotification.read
+      ? updateNotification.read
+      : oldNotification.read;
+    const link = updateNotification.link
+      ? updateNotification.link
+      : oldNotification.link;
```

Comment:
lol the best I can come up with is
```
const message = updateNotification.message ?? oldNotification.message;
const read = updateNotification.read ?? oldNotification.read;
const link = updateNotification.link ?? oldNotification.link;
```

Reply:
the other idea is returning 
```
return {
    ...oldNotification.
    ...updateNotification
};
```
which shoooould give the fields of oldNotification overwritten by whichever fields of updateNotification exist...?

Reply:
I swear I tried this and it was giving me the wrong result. Maybe I remembered wrong cuz that's why I didn't do it this way